File: packages/scheduling-engine/src/overflow/tier2-tardiness.ts
// ═══════════════════════════════════════════════════════════
//  INCOMPOL PLAN — Tier 2: Resolve TARDINESS
//  Blocks scheduled after their EDD. Uses advance + alt
//  machine. Constraint: never re-introduce overflow.
// ═══════════════════════════════════════════════════════════

import { MAX_AUTO_MOVES } from '../constants.js';
import type { AdvanceAction } from '../types/blocks.js';
import { computeAdvancedEdd, computeTardiness, sumOverflow } from './overflow-helpers.js';
import type { TierContext, TierState } from './tier-types.js';

/**
 * Tier 2: resolve tardiness (blocks scheduled after their EDD).
 * Mutates `state` in place. Returns the final tardiness value.
 */
export function runTier2(
  state: TierState,
  ctx: TierContext,
  preTier1TardyOps: Set<string>,
): void {
  const { userMoves, mSt, workdays, twinPartnerMap, runSchedule } = ctx;

  // Refresh tardy set: include ops that became tardy after Tier 1
  // (schedule shifts from overflow routing, twin co-production, etc.)
  for (const b of state.blocks) {
    if (b.type === 'ok' && b.eddDay != null && b.dayIdx > b.eddDay) {
      preTier1TardyOps.add(b.opId);
    }
  }

  let totalTardiness = computeTardiness(state.blocks);

  for (let t2 = 0; t2 < MAX_AUTO_MOVES && totalTardiness > 0; t2++) {
    if (state.autoMoves.length + state.autoAdvances.length >= MAX_AUTO_MOVES) break;

    // Separate moved vs advanced sets. Moved ops are fully excluded
    // (can't re-move). Advanced-but-still-tardy ops remain eligible
    // for Phase B/C (move to alt machine).
    const movedIds = new Set([
      ...userMoves.map((m) => m.opId),
      ...state.autoMoves.map((m) => m.opId),
    ]);
    const advancedIds = new Set(state.autoAdvances.map((a) => a.opId));

    const tardyBlocks = state.blocks
      .filter(
        (b) =>
          b.type === 'ok' &&
          b.eddDay != null &&
          b.dayIdx > b.eddDay &&
          preTier1TardyOps.has(b.opId) &&
          !movedIds.has(b.opId),
      )
      .sort((a, b) => b.prodMin - a.prodMin);

    if (tardyBlocks.length === 0) break;

    let tardyImproved = false;

    // ── Phase A: Try ADVANCING tardy ops on same machine ─────────
    // Skip ops already advanced (Phase A already tried them).
    const seenTardyOps = new Set<string>();
    for (const ob of tardyBlocks) {
      if (seenTardyOps.has(ob.opId)) continue;
      seenTardyOps.add(ob.opId);
      if (advancedIds.has(ob.opId)) continue; // already advanced, skip to Phase B/C

      const obEdd = ob.eddDay!;

      for (let advDays = 1; advDays <= 30; advDays++) {
        const targetDay = computeAdvancedEdd(obEdd, advDays, workdays);
        if (targetDay < 0) break;

        const trial: AdvanceAction[] = [
          ...state.autoAdvances,
          { opId: ob.opId, advanceDays: advDays, originalEdd: obEdd },
        ];
        const newResult = runSchedule([...userMoves, ...state.autoMoves], trial);
        const newTardiness = computeTardiness(newResult.blocks);

        if (newTardiness < totalTardiness && sumOverflow(newResult.blocks) === 0) {
          state.autoAdvances.push({ opId: ob.opId, advanceDays: advDays, originalEdd: obEdd });
          state.blocks = newResult.blocks;
          state.schedResult = newResult;
          totalTardiness = newTardiness;
          tardyImproved = true;

          // Mark blocks as advanced
          for (const b of state.blocks) {
            if (b.opId === ob.opId && b.type === 'ok') {
              b.isAdvanced = true;
              b.advancedByDays = advDays;
            }
          }

          // Record the advance production decision (tardy resolution)
          state.schedResult.registry.record({
            type: 'ADVANCE_PRODUCTION',
            opId: ob.opId,
            toolId: ob.toolId,
            machineId: ob.machineId,
            detail: `Advanced ${ob.sku} by ${advDays}d on ${ob.machineId} (tardy: EDD ${obEdd} -> ${targetDay})`,
            metadata: {
              originalEdd: obEdd,
              newEdd: targetDay,
              advanceDays: advDays,
              sku: ob.sku,
              reason: 'tardiness',
            },
          });

          break; // success with this advance amount
        }
      }

      if (tardyImproved) break; // restart Tier 2 loop with updated state
    }

    // Phase B: Alt machine for tardy ops.
    // The sumOverflow(nr.blocks) === 0 check inside ensures moves never
    // re-introduce overflow, so this is safe even after Tier 1 ran.
    if (!tardyImproved) {
      // Group tardy ops by alt machine — try batch move first (all ops sharing
      // the same alt), then fall back to individual moves. Batch moves let the
      // scheduler consolidate tool groups (fewer setups) on the alt machine.
      const altGroups = new Map<string, string[]>();
      const seenBatch = new Set<string>();
      for (const ob of tardyBlocks) {
        if (seenBatch.has(ob.opId)) continue;
        seenBatch.add(ob.opId);
        if (!ob.hasAlt || !ob.altM || mSt[ob.altM] === 'down') continue;
        const altM = ob.altM;
        if (!altGroups.has(altM)) altGroups.set(altM, []);
        altGroups.get(altM)!.push(ob.opId);
      }

      for (const [altM, opIds] of altGroups) {
        // Twin-aware: include twin partners in the batch to keep pairs together
        const batchOpIds = new Set(opIds);
        for (const opId of opIds) {
          const tp = twinPartnerMap.get(opId);
          if (tp && !movedIds.has(tp)) batchOpIds.add(tp);
        }
        const expandedOpIds = [...batchOpIds];

        // Try batch move (all ops to same alt at once)
        if (expandedOpIds.length > 1) {
          const batchMoves = expandedOpIds.map((opId) => ({ opId, toM: altM }));
          state.autoMoves.push(...batchMoves);
          const nr = runSchedule(
            [...userMoves, ...state.autoMoves],
            state.autoAdvances.length > 0 ? state.autoAdvances : undefined,
          );
          const nt = computeTardiness(nr.blocks);
          if (nt < totalTardiness && sumOverflow(nr.blocks) === 0) {
            state.blocks = nr.blocks;
            state.schedResult = nr;
            totalTardiness = nt;
            tardyImproved = true;
            for (const opId of opIds) {
              const blk = tardyBlocks.find((b) => b.opId === opId);
              if (blk) {
                state.schedResult.registry.record({
                  type: 'OVERFLOW_ROUTE',
                  opId,
                  toolId: blk.toolId,
                  machineId: altM,
                  detail: `Moved ${blk.sku} from ${blk.machineId} to ${altM} (tardy batch)`,
                  metadata: {
                    fromMachine: blk.machineId,
                    toMachine: altM,
                    sku: blk.sku,
                    reason: 'tardiness',
                  },
                });
              }
            }
            break;
          } else {
            state.autoMoves.splice(state.autoMoves.length - batchMoves.length);
          }
        }

        // Fall back to individual moves (try each op one at a time)
        if (!tardyImproved) {
          for (const opId of opIds) {
            // Twin-aware: also move twin partner to keep pair together
            const tp = twinPartnerMap.get(opId);
            const tpAlreadyMoved = tp ? movedIds.has(tp) || opIds.includes(tp) : true;
            state.autoMoves.push({ opId, toM: altM });
            if (tp && !tpAlreadyMoved) state.autoMoves.push({ opId: tp, toM: altM });
            const nr = runSchedule(
              [...userMoves, ...state.autoMoves],
              state.autoAdvances.length > 0 ? state.autoAdvances : undefined,
            );
            const nt = computeTardiness(nr.blocks);
            if (nt < totalTardiness && sumOverflow(nr.blocks) === 0) {
              state.blocks = nr.blocks;
              state.schedResult = nr;
              totalTardiness = nt;
              tardyImproved = true;
              const blk = tardyBlocks.find((b) => b.opId === opId);
              if (blk) {
                state.schedResult.registry.record({
                  type: 'OVERFLOW_ROUTE',
                  opId,
                  toolId: blk.toolId,
                  machineId: altM,
                  detail: `Moved ${blk.sku} from ${blk.machineId} to ${altM} (tardy resolution)`,
                  metadata: {
                    fromMachine: blk.machineId,
                    toMachine: altM,
                    sku: blk.sku,
                    reason: 'tardiness',
                  },
                });
              }
              break;
            } else {
              state.autoMoves.pop();
              if (tp && !tpAlreadyMoved) state.autoMoves.pop();
            }
          }
        }

        if (tardyImproved) break;
      }
    }

    // ── Phase C: Combined MOVE + ADVANCE for tardy ops ──────────
    // If Phase A (advance-only) and Phase B (move-only) both failed,
    // try moving the op to its alt machine AND advancing it by N days.
    // This lets the scheduler use earlier (less congested) slots on
    // the alt machine — solving cases where the alt is full near EDD
    // but has capacity earlier.
    if (!tardyImproved) {
      const seenC = new Set<string>();
      for (const ob of tardyBlocks) {
        if (seenC.has(ob.opId)) continue;
        seenC.add(ob.opId);
        if (!ob.hasAlt || !ob.altM || mSt[ob.altM] === 'down') continue;
        const altM = ob.altM;
        const obEdd = ob.eddDay!;

        // Twin-aware: move partner along with the primary op
        const tpC = twinPartnerMap.get(ob.opId);
        const tpCAlreadyMoved = tpC ? movedIds.has(tpC) : true;

        for (let advDays = 1; advDays <= 30; advDays++) {
          const targetDay = computeAdvancedEdd(obEdd, advDays, workdays);
          if (targetDay < 0) break;

          // Try move + advance together (include twin partner in move)
          state.autoMoves.push({ opId: ob.opId, toM: altM });
          if (tpC && !tpCAlreadyMoved) state.autoMoves.push({ opId: tpC, toM: altM });
          const trialAdv: AdvanceAction[] = [
            ...state.autoAdvances,
            { opId: ob.opId, advanceDays: advDays, originalEdd: obEdd },
          ];
          const nr = runSchedule([...userMoves, ...state.autoMoves], trialAdv);
          const nt = computeTardiness(nr.blocks);

          if (nt < totalTardiness && sumOverflow(nr.blocks) === 0) {
            state.autoAdvances.push({ opId: ob.opId, advanceDays: advDays, originalEdd: obEdd });
            state.blocks = nr.blocks;
            state.schedResult = nr;
            totalTardiness = nt;
            tardyImproved = true;

            // Mark blocks as advanced
            for (const b of state.blocks) {
              if (b.opId === ob.opId && b.type === 'ok') {
                b.isAdvanced = true;
                b.advancedByDays = advDays;
              }
            }

            state.schedResult.registry.record({
              type: 'OVERFLOW_ROUTE',
              opId: ob.opId,
              toolId: ob.toolId,
              machineId: altM,
              detail: `Moved ${ob.sku} from ${ob.machineId} to ${altM} + advanced ${advDays}d (tardy combo)`,
              metadata: {
                fromMachine: ob.machineId,
                toMachine: altM,
                sku: ob.sku,
                advanceDays: advDays,
                reason: 'tardiness',
              },
            });
            break;
          } else {
            state.autoMoves.pop(); // undo the trial move
            if (tpC && !tpCAlreadyMoved) state.autoMoves.pop();
          }
        }

        if (tardyImproved) break;
      }
    }

    // ── Phase D: BATCH advance all tardy ops by 1 day ───────
    // When individual advances fail (local minimum), try advancing
    // ALL tardy ops simultaneously. This helps when multiple ops
    // compete for the same capacity window — moving them ALL earlier
    // can break the deadlock that individual advances can't.
    if (!tardyImproved) {
      // Collect unique tardy ops (not yet moved)
      const batchOps = new Map<string, number>(); // opId → eddDay
      for (const ob of tardyBlocks) {
        if (!batchOps.has(ob.opId) && !advancedIds.has(ob.opId) && ob.eddDay != null) {
          batchOps.set(ob.opId, ob.eddDay);
        }
      }

      if (batchOps.size > 1) {
        for (let advDays = 1; advDays <= 5; advDays++) {
          const batchAdvances: AdvanceAction[] = [];
          for (const [opId, edd] of batchOps) {
            const targetDay = computeAdvancedEdd(edd, advDays, workdays);
            if (targetDay >= 0) {
              batchAdvances.push({ opId, advanceDays: advDays, originalEdd: edd });
            }
          }

          if (batchAdvances.length < 2) continue;

          const trial = [...state.autoAdvances, ...batchAdvances];
          const newResult = runSchedule([...userMoves, ...state.autoMoves], trial);
          const newTardiness = computeTardiness(newResult.blocks);

          if (newTardiness < totalTardiness && sumOverflow(newResult.blocks) === 0) {
            for (const ba of batchAdvances) state.autoAdvances.push(ba);
            state.blocks = newResult.blocks;
            state.schedResult = newResult;
            totalTardiness = newTardiness;
            tardyImproved = true;

            for (const b of state.blocks) {
              if (batchOps.has(b.opId) && b.type === 'ok') {
                b.isAdvanced = true;
                b.advancedByDays = advDays;
              }
            }
            break;
          }
        }
      }
    }

    if (!tardyImproved) break;
  }
}

File: apps/backend/src/domain/scheduling/overflow/tier2_tardiness.py
"""Tier 2: Resolve TARDINESS — port of overflow/tier2-tardiness.ts.

Blocks scheduled after their EDD. Uses advance + alt machine.
Constraint: never re-introduce overflow.
"""

from __future__ import annotations

from ..constants import MAX_AUTO_MOVES
from ..types import AdvanceAction
from .overflow_helpers import compute_advanced_edd, compute_tardiness, sum_overflow
from .tier_types import TierContext, TierState


def run_tier2(
    state: TierState,
    ctx: TierContext,
    pre_tier1_tardy_ops: set[str],
) -> None:
    """Tier 2: resolve tardiness (blocks scheduled after their EDD).

    Mutates `state` in place.
    """
    user_moves = ctx.user_moves
    m_st = ctx.m_st
    workdays = ctx.workdays
    twin_partner_map = ctx.twin_partner_map
    run_schedule = ctx.run_schedule

    # Refresh tardy set: include ops that became tardy after Tier 1
    for b in state.blocks:
        if b.type == "ok" and b.edd_day is not None and b.day_idx > b.edd_day:
            pre_tier1_tardy_ops.add(b.op_id)

    total_tardiness = compute_tardiness(state.blocks)

    for _t2 in range(MAX_AUTO_MOVES):
        if total_tardiness <= 0:
            break
        if len(state.auto_moves) + len(state.auto_advances) >= MAX_AUTO_MOVES:
            break

        moved_ids = set([m.op_id for m in user_moves] + [m.op_id for m in state.auto_moves])
        advanced_ids = set(a.op_id for a in state.auto_advances)

        tardy_blocks = sorted(
            [
                b
                for b in state.blocks
                if b.type == "ok"
                and b.edd_day is not None
                and b.day_idx > b.edd_day
                and b.op_id in pre_tier1_tardy_ops
                and b.op_id not in moved_ids
            ],
            key=lambda b: b.prod_min,
            reverse=True,
        )

        if not tardy_blocks:
            break

        tardy_improved = False

        # ── Phase A: Try ADVANCING tardy ops on same machine ──
        seen_tardy_ops: set[str] = set()
        for ob in tardy_blocks:
            if ob.op_id in seen_tardy_ops:
                continue
            seen_tardy_ops.add(ob.op_id)
            if ob.op_id in advanced_ids:
                continue  # already advanced, skip to Phase B/C

            ob_edd = ob.edd_day  # type: ignore[assignment]

            for adv_days in range(1, 31):
                target_day = compute_advanced_edd(ob_edd, adv_days, workdays)
                if target_day < 0:
                    break

                trial = list(state.auto_advances) + [
                    AdvanceAction(op_id=ob.op_id, advance_days=adv_days, original_edd=ob_edd)
                ]
                all_moves = list(user_moves) + list(state.auto_moves)
                new_result = run_schedule(all_moves, trial)
                new_tardiness = compute_tardiness(new_result.blocks)

                if new_tardiness < total_tardiness and sum_overflow(new_result.blocks) == 0:
                    state.auto_advances.append(
                        AdvanceAction(op_id=ob.op_id, advance_days=adv_days, original_edd=ob_edd)
                    )
                    state.blocks = new_result.blocks
                    state.sched_result = new_result
                    total_tardiness = new_tardiness
                    tardy_improved = True

                    # Mark blocks as advanced
                    for b in state.blocks:
                        if b.op_id == ob.op_id and b.type == "ok":
                            b.is_advanced = True
                            b.advanced_by_days = adv_days

                    break  # success

            if tardy_improved:
                break  # restart loop

        # ── Phase B: Alt machine for tardy ops ──
        if not tardy_improved:
            alt_groups: dict[str, list[str]] = {}
            seen_batch: set[str] = set()
            for ob in tardy_blocks:
                if ob.op_id in seen_batch:
                    continue
                seen_batch.add(ob.op_id)
                if not ob.has_alt or not ob.alt_m or m_st.get(ob.alt_m) == "down":
                    continue
                alt_m = ob.alt_m
                if alt_m not in alt_groups:
                    alt_groups[alt_m] = []
                alt_groups[alt_m].append(ob.op_id)

            for alt_m, op_ids in alt_groups.items():
                # Twin-aware: include twin partners
                batch_op_ids = set(op_ids)
                for op_id in op_ids:
                    tp = twin_partner_map.get(op_id)
                    if tp and tp not in moved_ids:
                        batch_op_ids.add(tp)
                expanded_op_ids = list(batch_op_ids)

                # Try batch move
                if len(expanded_op_ids) > 1:
                    from ..types import MoveAction as MA

                    batch_moves = [MA(op_id=oid, to_m=alt_m) for oid in expanded_op_ids]
                    state.auto_moves.extend(batch_moves)
                    all_moves = list(user_moves) + list(state.auto_moves)
                    adv = state.auto_advances if state.auto_advances else None
                    nr = run_schedule(all_moves, adv)
                    nt = compute_tardiness(nr.blocks)
                    if nt < total_tardiness and sum_overflow(nr.blocks) == 0:
                        state.blocks = nr.blocks
                        state.sched_result = nr
                        total_tardiness = nt
                        tardy_improved = True
                    else:
                        del state.auto_moves[-len(batch_moves) :]

                # Fall back to individual moves
                if not tardy_improved:
                    for op_id in op_ids:
                        tp = twin_partner_map.get(op_id)
                        tp_already_moved = tp in moved_ids or tp in op_ids if tp else True
                        state.auto_moves.append(MA(op_id=op_id, to_m=alt_m))
                        if tp and not tp_already_moved:
                            state.auto_moves.append(MA(op_id=tp, to_m=alt_m))
                        all_moves = list(user_moves) + list(state.auto_moves)
                        adv = state.auto_advances if state.auto_advances else None
                        nr = run_schedule(all_moves, adv)
                        nt = compute_tardiness(nr.blocks)
                        if nt < total_tardiness and sum_overflow(nr.blocks) == 0:
                            state.blocks = nr.blocks
                            state.sched_result = nr
                            total_tardiness = nt
                            tardy_improved = True
                            break
                        else:
                            state.auto_moves.pop()
                            if tp and not tp_already_moved:
                                state.auto_moves.pop()

                if tardy_improved:
                    break

        # ── Phase C: Combined MOVE + ADVANCE ──
        if not tardy_improved:
            seen_c: set[str] = set()
            for ob in tardy_blocks:
                if ob.op_id in seen_c:
                    continue
                seen_c.add(ob.op_id)
                if not ob.has_alt or not ob.alt_m or m_st.get(ob.alt_m) == "down":
                    continue
                alt_m = ob.alt_m
                ob_edd = ob.edd_day  # type: ignore[assignment]

                tp_c = twin_partner_map.get(ob.op_id)
                tp_c_already_moved = tp_c in moved_ids if tp_c else True

                from ..types import MoveAction as MA

                for adv_days in range(1, 31):
                    target_day = compute_advanced_edd(ob_edd, adv_days, workdays)
                    if target_day < 0:
                        break

                    state.auto_moves.append(MA(op_id=ob.op_id, to_m=alt_m))
                    if tp_c and not tp_c_already_moved:
                        state.auto_moves.append(MA(op_id=tp_c, to_m=alt_m))
                    trial_adv = list(state.auto_advances) + [
                        AdvanceAction(op_id=ob.op_id, advance_days=adv_days, original_edd=ob_edd)
                    ]
                    all_moves = list(user_moves) + list(state.auto_moves)
                    nr = run_schedule(all_moves, trial_adv)
                    nt = compute_tardiness(nr.blocks)

                    if nt < total_tardiness and sum_overflow(nr.blocks) == 0:
                        state.auto_advances.append(
                            AdvanceAction(
                                op_id=ob.op_id, advance_days=adv_days, original_edd=ob_edd
                            )
                        )
                        state.blocks = nr.blocks
                        state.sched_result = nr
                        total_tardiness = nt
                        tardy_improved = True

                        for b in state.blocks:
                            if b.op_id == ob.op_id and b.type == "ok":
                                b.is_advanced = True
                                b.advanced_by_days = adv_days
                        break
                    else:
                        state.auto_moves.pop()
                        if tp_c and not tp_c_already_moved:
                            state.auto_moves.pop()

                if tardy_improved:
                    break

        # ── Phase D: BATCH advance all tardy ops ──
        if not tardy_improved:
            batch_ops: dict[str, int] = {}
            for ob in tardy_blocks:
                if (
                    ob.op_id not in batch_ops
                    and ob.op_id not in advanced_ids
                    and ob.edd_day is not None
                ):
                    batch_ops[ob.op_id] = ob.edd_day

            if len(batch_ops) > 1:
                for adv_days in range(1, 6):
                    batch_advances: list[AdvanceAction] = []
                    for op_id, edd in batch_ops.items():
                        target_day = compute_advanced_edd(edd, adv_days, workdays)
                        if target_day >= 0:
                            batch_advances.append(
                                AdvanceAction(op_id=op_id, advance_days=adv_days, original_edd=edd)
                            )

                    if len(batch_advances) < 2:
                        continue

                    trial = list(state.auto_advances) + batch_advances
                    all_moves = list(user_moves) + list(state.auto_moves)
                    new_result = run_schedule(all_moves, trial)
                    new_tardiness = compute_tardiness(new_result.blocks)

                    if new_tardiness < total_tardiness and sum_overflow(new_result.blocks) == 0:
                        state.auto_advances.extend(batch_advances)
                        state.blocks = new_result.blocks
                        state.sched_result = new_result
                        total_tardiness = new_tardiness
                        tardy_improved = True

                        for b in state.blocks:
                            if b.op_id in batch_ops and b.type == "ok":
                                b.is_advanced = True
                                b.advanced_by_days = adv_days
                        break

        if not tardy_improved:
            break
